### PR TITLE
isInDirectory(): handle trailing slashes

### PR DESCRIPTION
**Pull request by justinmk3**
If directory has trailing slashes, they do not change the logical question of whether some other path is a descendant of the directory.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by justinmk3** [comment]
This will require too much churn. Better to merge feature/debugconfig to master soon.

File: src/test/shared/filesystemUtilities.test.ts [hunk 1]
```diff
@@ -101,33 +101,17 @@ describe('filesystemUtilities', () => {
...
-        it('returns false for a similarly named dir', () => {
-            assert.ok(!isContainedWithinDirectory(basePath, `${basePath}point`))
-        })
+        assert.ok(isInDirectory(basePath, basePath))
+        assert.ok(isInDirectory(basePath, extendedPath))
+        assert.ok(isInDirectory(basePath, path.join(basePath, filename)))
+        assert.ok(isInDirectory(basePath, path.join(extendedPath, filename)))
+        assert.ok(!isInDirectory(basePath, path.join('what', 'are', 'you', 'looking', 'at')))
+        assert.ok(!isInDirectory(basePath, `${basePath}point`))
+        assert.ok(isInDirectory('/foo/bar/baz/', '/foo/bar/baz/a.txt'))
```
Comment (justinmk3) [inline src/test/shared/filesystemUtilities.test.ts:115]: This test fails before the fix.

File: src/test/shared/filesystemUtilities.test.ts [hunk 2]
```diff
@@ -101,33 +101,17 @@ describe('filesystemUtilities', () => {
...
-
-        it('returns false for a similarly named dir', () => {
-            assert.ok(!isContainedWithinDirectory(basePath, `${basePath}point`))
-        })
+        assert.ok(isInDirectory(basePath, basePath))
+        assert.ok(isInDirectory(basePath, extendedPath))
+        assert.ok(isInDirectory(basePath, path.join(basePath, filename)))
+        assert.ok(isInDirectory(basePath, path.join(extendedPath, filename)))
+        assert.ok(!isInDirectory(basePath, path.join('what', 'are', 'you', 'looking', 'at')))
+        assert.ok(!isInDirectory(basePath, `${basePath}point`))
```
Comment (justinmk3) [inline src/test/shared/filesystemUtilities.test.ts:114]: this is a much more compact way to write these tests without any loss of clarity. The old test granular descriptions aren't helpful: they struggle to find a purpose, and that's a clue that they aren't useful.

Comment (awschristou) [reply src/test/shared/filesystemUtilities.test.ts:114]: The value the former approach takes is that you have a full (and in theory, descriptive) list of which scenarios break down when there is a failure. The changed approach fails on first issue, without any high level description (think CI reports) as to what situation is broken, and which scenarios are broken.

A parameterized test is the best fit, a crude version of which can be seen in some places in the testing code.

File: src/shared/filesystemUtilities.ts [hunk 3]
```diff
@@ -78,14 +78,22 @@ export const makeTemporaryToolkitFolder = async (...relativePathParts: string[])
...
-    const containedPathPieces = containedPath.split(path.sep)
+export function isInDirectory(d: string, p: string): boolean {
+    const parentDirPieces = d.split(path.sep)
+    const containedPathPieces = p.split(path.sep)
+    // Remove final empty element(s), if `d` ends with slash(es).
+    while (parentDirPieces.length > 0 && parentDirPieces[parentDirPieces.length - 1] === '') {
+        parentDirPieces.pop()
+    }
 
     return parentDirPieces.every((value, index) => {
```
Comment (awschristou) [inline src/shared/filesystemUtilities.ts:98]: If you've popped all of parentDirPieces, isn't it empty now? Does this give false positives?

Comment (justinmk3) [reply src/shared/filesystemUtilities.ts:98]: > If you've popped all of parentDirPieces, isn't it empty now?

That would happen if the directory is `/` or `/////` , which logically contains "everything". Is there some other scenario you are thinking of?

Comment (awschristou) [reply src/shared/filesystemUtilities.ts:98]: Ah I see. Thanks, I misunderstood the code.

Something that is not related to your change, but I just realized from this nth look at the method, we're splitting on `path.sep`, what happens (for example) on windows, when you feed this a "normalized path that uses the / slash?

File: src/test/shared/filesystemUtilities.test.ts (hunk 2 above)
Comment (justinmk3) [reply src/test/shared/filesystemUtilities.test.ts:114]: > you have a full (and in theory, descriptive) list of which scenarios break down when there is a failure.

But failure is the uncommon case, and the gain is very small: if this test fails in the new form, it is easy to see the problem. If the next case fails, well, that's a _very uncommon_ scenario, and in any case the cost is an extra test run. That cost vs having verbose, low-density test cases is a questionable tradeoff.

We should optimize for the common case.

> A parameterized test is the best fit, a crude version of which can be seen in some places in the testing code.

Those tests tend to be even harder to read, because the codepath is not lexical. These literal cases are obvious and easy to read.

File: src/shared/filesystemUtilities.ts (hunk 3 above)
Comment (justinmk3) [reply src/shared/filesystemUtilities.ts:98]: yeah we should probably normalize here. That's why windows CI is failing...

Comment (awschristou) [reply src/shared/filesystemUtilities.ts:98]: Lucky that there is a test using explicit "/" slashes. We should probably add a test that uses explicit "\" slashes to exercise appropriately on the non-windows systems too (though that is a far more uncommon scenario)

File: src/test/shared/filesystemUtilities.test.ts (hunk 2 above)
Comment (awschristou) [reply src/test/shared/filesystemUtilities.test.ts:114]: > But failure is the uncommon case, and the gain is very small: if this test fails in the new form, it is easy to see the problem.

I am not comfortable with the spirit of that statement (for any tests, not as it applies to this method being tested). When edge cases reveal a failure, it isn't necessarily easy to see the problem, or which scenarios are affected. You only know that _something_ is wrong with foo(). Getting a result that says "expected: true, actual: false" doesn't get you any closer to knowing where the issue is, which requires more effort to triage.

Comment (justinmk3) [reply src/test/shared/filesystemUtilities.test.ts:115]: It is not a general principle, it is context-sensitive. 

In fact most of the suggestions I find myself advocating fall under the general idea that we should use more discerning practices.  We should not be hyper-granular just out of habit. I'm 100% in favor of descriptive test cases for more complicated functions. `isInDirectory()` is an isolated piece of logic, which may take infinite inputs. Having infinite test cases to model those inputs makes tests harder to read and harder to write, and optimizes for a very uncommon scenario.

File: src/test/shared/filesystemUtilities.test.ts (hunk 1 above)
Comment (awschristou) [reply src/test/shared/filesystemUtilities.test.ts:116]: My comment earlier (https://github.com/aws/aws-toolkit-vscode/pull/1036/files/e7230ff600e3a1e914f1a17c7bdc6d328d6ad89f..b4f8d1d37bd1cd5b9c867290b0257de57f26ae5f#r409049401) failed to render correctly.  Should there be a test that uses the explicit `\` slash in pathing as well?

File: src/shared/filesystemUtilities.ts [hunk 4]
```diff
@@ -78,16 +79,30 @@ export const makeTemporaryToolkitFolder = async (...relativePathParts: string[])
...
+ *
+ * @param d  Path to a directory.
+ * @param p  Path to file or directory to test.
  */
-export function isContainedWithinDirectory(parentDirectory: string, containedPath: string): boolean {
-    const parentDirPieces = parentDirectory.split(path.sep)
-    const containedPathPieces = containedPath.split(path.sep)
+export function isInDirectory(d: string, p: string): boolean {
+    if (d === '' || p === '') {
+        return true
```
Comment (justinmk3) [inline src/shared/filesystemUtilities.ts:93]: Explicit decision for this edge case. Covered by new tests.

File: src/shared/filesystemUtilities.ts [hunk 5]
```diff
@@ -78,16 +79,30 @@ export const makeTemporaryToolkitFolder = async (...relativePathParts: string[])
...
+    while (parentDirPieces.length > 0 && parentDirPieces[parentDirPieces.length - 1] === '') {
+        parentDirPieces.pop()
+    }
+    const caseInsensitive = os.platform() === 'win32'
 
     return parentDirPieces.every((value, index) => {
-        return value === containedPathPieces[index]
+        return caseInsensitive
+            ? value.toLowerCase() === containedPathPieces[index].toLowerCase()
+            : value === containedPathPieces[index]
```
Comment (justinmk3) [inline src/shared/filesystemUtilities.ts:106]: handle case-insensitivity

File: src/test/shared/filesystemUtilities.test.ts (hunk 1 above)
Comment (justinmk3) [reply src/test/shared/filesystemUtilities.test.ts:116]: done
